### PR TITLE
nceplibs

### LINKED ISSUE #43: Improve build of NCEPlibs
Currently we install NCEPlibs from a subdirectory in jedi-stack because its convenient.  But, we should really download each library from the github site: https://github.com/NOAA-EMC and then configure each one as described there.  But, I'm not sure it's worth the trouble now because (1) JEDI doesn't currently need them and (2) EMC is working on a new CMake build system for them.  Still, I'm putting it in the icebox as something that should be done eventually.

### PR DESCRIPTION
**Pull request by aerorahul**
Use the NOAA-EMC NCEPLIBS umbrella repository to obtain and build NCEP libraries.

closes #43

### CODE REVIEW

File: buildscripts/libs/build_nceplibs.sh [hunk 1]
```diff
@@ -1,52 +1,59 @@
...
 
-if [[ $(echo $compiler | cut -d- -f1) = "gnu" ]]; then
-    ln -s macros.make.cheyenne.gnu macros.make
-elif [[ $(echo $compiler | cut -d- -f1) = "intel" ]]; then
-    ln -s macros.make.aws.intel macros.make
-fi
-make
+export FFLAGS="-fPIC"
+export CFLAGS="-fPIC"
+export CXXFLAGS="-fPIC"
```
Comment (markjolah) [inline buildscripts/libs/build_nceplibs.sh:40]: Can these be `+=` to append to already existing flags?

File: buildscripts/libs/build_nceplibs.sh [hunk 2]
```diff
@@ -1,52 +1,59 @@
...
+mkdir -p build && cd build
 
-# generate modulefile from template
-$MODULES && update_modules compiler $name $version \
-         || echo $name $version >> ${JEDI_STACK_ROOT}/jedi-stack-contents.log
+cmake -DCMAKE_INSTALL_PREFIX=$prefix \
+      -DDEPLOY=ON \
+      -DBUILD_CRTM=OFF \
+      -DBUILD_POST=OFF ..
+make -j${NTHREADS:-4}
```
Comment (markjolah) [inline buildscripts/libs/build_nceplibs.sh:58]: Enable verbose output for debugging `VERBOSE=$MAKE_VERBOSE make -j${NTHREADS:-4}`

File: buildscripts/libs/build_nceplibs.sh [hunk 3]
```diff
@@ -1,52 +1,59 @@
...
+make deploy
```
Comment (markjolah) [inline buildscripts/libs/build_nceplibs.sh:59]: You may need a $SUDO here if you are ever going to run this on the OSX?

File: buildscripts/libs/build_nceplibs.sh (hunk 2 above)
Comment (aerorahul) [reply buildscripts/libs/build_nceplibs.sh]: done.

File: buildscripts/libs/build_nceplibs.sh (hunk 3 above)
Comment (aerorahul) [reply buildscripts/libs/build_nceplibs.sh]: unclear, but I added it anyways.

File: buildscripts/libs/build_nceplibs.sh (hunk 1 above)
Comment (aerorahul) [reply buildscripts/libs/build_nceplibs.sh]: actually, on second thought. I don't think this is a good idea.
The reason these scripts have these flags is precisely to not use anything from environment or elsewhere.

**Comment by mmiesch** [comment]
Wow - that issue was opened a year ago but I wasn't sure how to implement it - thanks @aerorahul!  Overwhelmed with other stuff today but I'll test it when I get a chance.

File: buildscripts/libs/build_nceplibs.sh (hunk 1 above)
Comment (mmiesch) [reply buildscripts/libs/build_nceplibs.sh]: Yes - I agree with @aerorahul - this was deliberate to isolate the flags needed for a particular package

**Comment by markjolah** [comment]
I'm ok with merging this if it is working for @aerorahul's purposes.  I cannot test this with intel until we resolve the other #120 re compiler flags.  I suppose I could try on OSX with clang/gfortran.

**Comment by mmiesch** [comment]
After fixing the github URL, it builds ok but is failing on the install:
```bash
-- Installing: /opt/modules/gnu-9.3.0/openmpi-4.0.3/bufr-11.3.1.1/lib/libbufr_d_DA.a
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/ranlib: file: /opt/modules/gnu-9.3.0/openmpi-4.0.3/bufr-11.3.1.1/lib/libbufr_d_DA.a(modv_MXH4WLC.F.o) has no symbols
-- Installing: /opt/modules/gnu-9.3.0/openmpi-4.0.3/bufr-11.3.1.1/lib/cmake/bufr/bufr-config.cmake
-- Installing: /opt/modules/gnu-9.3.0/openmpi-4.0.3/bufr-11.3.1.1/lib/cmake/bufr/bufr-config-version.cmake
-- Installing: /opt/modules/gnu-9.3.0/openmpi-4.0.3/bufr-11.3.1.1/lib/cmake/bufr/bufr-targets.cmake
-- Installing: /opt/modules/gnu-9.3.0/openmpi-4.0.3/bufr-11.3.1.1/lib/cmake/bufr/bufr-targets-release.cmake
cd /Users/miesch/JEDI/code/local_copy/jedi-stack/pkg/nceplibs-develop/build/bufr/src/bufr-build && /usr/local/Cellar/cmake/3.17.2/bin/cmake -E touch /Users/miesch/JEDI/code/local_copy/jedi-stack/pkg/nceplibs-develop/build/bufr/src/bufr-stamp/bufr-install
[ 57%] Completed 'bufr'
/usr/local/Cellar/cmake/3.17.2/bin/cmake -E make_directory /Users/miesch/JEDI/code/local_copy/jedi-stack/pkg/nceplibs-develop/build/CMakeFiles
/usr/local/Cellar/cmake/3.17.2/bin/cmake -E touch /Users/miesch/JEDI/code/local_copy/jedi-stack/pkg/nceplibs-develop/build/CMakeFiles/bufr-complete
/usr/local/Cellar/cmake/3.17.2/bin/cmake -E touch /Users/miesch/JEDI/code/local_copy/jedi-stack/pkg/nceplibs-develop/build/bufr/src/bufr-stamp/bufr-done
[ 57%] Built target bufr
make: *** [all] Error 2
BUILD FAIL!  Lib: nceplibs-develop Error:2
```

This is for Mac OS with gnu/openmpi

File: buildscripts/build_stack.sh [hunk 4]
```diff
@@ -103,7 +103,7 @@ build_lib ODC odc jcsda 1.0.3
 build_lib JASPER jasper 1.900.1
 build_lib ARMADILLO armadillo 1.900.1
 build_lib XERCES xerces 3.1.4
-build_lib NCEPLIBS nceplibs fv3
+build_lib NCEPLIBS nceplibs develop
```
Comment (mmiesch) [inline buildscripts/build_stack.sh:106]: Do we want to use develop here or is there a tagged version we should use for stability?

Comment (aerorahul) [reply buildscripts/build_stack.sh:106]: develop is fine for now.  I don't think there is a tag. 
NCEPLIBS is an umbrella repository that contains pointers to all the individual ones.
At some point, I think I will turn `build_nceplibs.sh` into a loop over all the individual ones to control the exact version of the subcomponents.

**Comment by aerorahul** [comment]
> After fixing the github URL, it builds ok but is failing on the install:
> 
> ```shell
> -- Installing: /opt/modules/gnu-9.3.0/openmpi-4.0.3/bufr-11.3.1.1/lib/libbufr_d_DA.a
> /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/ranlib: file: /opt/modules/gnu-9.3.0/openmpi-4.0.3/bufr-11.3.1.1/lib/libbufr_d_DA.a(modv_MXH4WLC.F.o) has no symbols
> -- Installing: /opt/modules/gnu-9.3.0/openmpi-4.0.3/bufr-11.3.1.1/lib/cmake/bufr/bufr-config.cmake
> -- Installing: /opt/modules/gnu-9.3.0/openmpi-4.0.3/bufr-11.3.1.1/lib/cmake/bufr/bufr-config-version.cmake
> -- Installing: /opt/modules/gnu-9.3.0/openmpi-4.0.3/bufr-11.3.1.1/lib/cmake/bufr/bufr-targets.cmake
> -- Installing: /opt/modules/gnu-9.3.0/openmpi-4.0.3/bufr-11.3.1.1/lib/cmake/bufr/bufr-targets-release.cmake
> cd /Users/miesch/JEDI/code/local_copy/jedi-stack/pkg/nceplibs-develop/build/bufr/src/bufr-build && /usr/local/Cellar/cmake/3.17.2/bin/cmake -E touch /Users/miesch/JEDI/code/local_copy/jedi-stack/pkg/nceplibs-develop/build/bufr/src/bufr-stamp/bufr-install
> [ 57%] Completed 'bufr'
> /usr/local/Cellar/cmake/3.17.2/bin/cmake -E make_directory /Users/miesch/JEDI/code/local_copy/jedi-stack/pkg/nceplibs-develop/build/CMakeFiles
> /usr/local/Cellar/cmake/3.17.2/bin/cmake -E touch /Users/miesch/JEDI/code/local_copy/jedi-stack/pkg/nceplibs-develop/build/CMakeFiles/bufr-complete
> /usr/local/Cellar/cmake/3.17.2/bin/cmake -E touch /Users/miesch/JEDI/code/local_copy/jedi-stack/pkg/nceplibs-develop/build/bufr/src/bufr-stamp/bufr-done
> [ 57%] Built target bufr
> make: *** [all] Error 2
> BUILD FAIL!  Lib: nceplibs-develop Error:2
> ```
> 
> This is for Mac OS with gnu/openmpi

Interesting. I don't get that error on macOS with gnu/openmpi 9.3.0/3.1.5

**Comment by mmiesch** [comment]
I'm also having trouble on AWS with intel.  Strange.  At first Cmake seems to find the compiler fine:
```bash
+ cmake -DCMAKE_INSTALL_PREFIX=/optjedi/modules/intel-20.0.166/impi-20.0.166 -DDEPLOY=ON -DBUILD_CRTM=OFF -DBUILD_POST=OFF ..
-- The C compiler identification is Intel 19.1.0.20191121
-- The Fortran compiler identification is Intel 19.1.0.20191121
```

But then during the actual make it gets confused:
```bash
+ sudo VERBOSE=Y make -j4
[...]
Re-run cmake no build system arguments
[ 14%] Performing configure step for 'bacio'
cd /home/ubuntu/jedi/jedi-stack/pkg/nceplibs-develop/build/bacio/src/bacio-build && /optjedi/modules/core/cmake/3.17.3/bin/cmake "-DCMAKE_PREFIX_PATH=/optjedi/modules/intel-20.0.166/impi-20.0.166;/optjedi/modules/intel-20.0.166/impi-20.0.166/bacio;/optjedi/modules/intel-20.0.166/impi-20.0.166/sigio;/optjedi/modules/intel-20.0.166/impi-20.0.166/sfcio;/optjedi/modules/intel-20.0.166/impi-20.0.166/gfsio;/optjedi/modules/intel-20.0.166/impi-20.0.166/ip;/optjedi/modules/intel-20.0.166/impi-20.0.166/sp;/optjedi/modules/intel-20.0.166/impi-20.0.166/w3nco;/optjedi/modules/intel-20.0.166/impi-20.0.166/g2;/optjedi/modules/intel-20.0.166/impi-20.0.166/g2tmpl;/optjedi/modules/intel-20.0.166/impi-20.0.166/nemsio;/optjedi/modules/intel-20.0.166/impi-20.0.166/nemsiogfs;/optjedi/modules/intel-20.0.166/impi-20.0.166/landsfcutil;/optjedi/modules/intel-20.0.166/impi-20.0.166/w3emc;/optjedi/modules/intel-20.0.166/impi-20.0.166/wrf_io;/optjedi/modules/intel-20.0.166/impi-20.0.166/bufr" -DCMAKE_INSTALL_PREFIX=/optjedi/modules/intel-20.0.166/impi-20.0.166/bacio-2.4.0 -DENABLE_TESTS=OFF -DCMAKE_C_COMPILER=/optjedi/intel/compilers_and_libraries_2020.0.166/linux/mpi/intel64/bin/mpiicc -DCMAKE_Fortran_COMPILER=/optjedi/intel/compilers_and_libraries_2020.0.166/linux/mpi/intel64/bin/mpiifort "-GUnix Makefiles" /home/ubuntu/jedi/jedi-stack/pkg/nceplibs-develop/build/bacio/src/bacio
Re-run cmake no build system arguments
-- The Fortran compiler identification is unknown
/optjedi/intel/compilers_and_libraries_2020.0.166/linux/mpi/intel64/bin/mpiifort: 1: eval: ifort: not found
-- Check for working Fortran compiler: /optjedi/intel/compilers_and_libraries_2020.0.166/linux/mpi/intel64/bin/mpiifort
-- The Fortran compiler identification is unknown
/optjedi/intel/compilers_and_libraries_2020.0.166/linux/mpi/intel64/bin/mpiifort: 1: eval: ifort: not found
-- Check for working Fortran compiler: /optjedi/intel/compilers_and_libraries_2020.0.166/linux/mpi/intel64/bin/mpiifort
-- Check for working Fortran compiler: /optjedi/intel/compilers_and_libraries_2020.0.166/linux/mpi/intel64/bin/mpiifort - broken
CMake Error at /optjedi/modules/core/cmake/3.17.3/share/cmake-3.17/Modules/CMakeTestFortranCompiler.cmake:45 (message):
  The Fortran compiler

    "/optjedi/intel/compilers_and_libraries_2020.0.166/linux/mpi/intel64/bin/mpiifort"

  is not able to compile a simple test program.

  It fails with the following output:

    Change Dir: /home/ubuntu/jedi/jedi-stack/pkg/nceplibs-develop/build/gfsio/src/gfsio-build/CMakeFiles/CMakeTmp
    
    Run Build Command(s):/usr/bin/make cmTC_7b38f/fast && make[3]: Entering directory '/home/ubuntu/jedi/jedi-stack/pkg/nceplibs-develop/build/gfsio/src/gfsio-build/CMakeFiles/CMakeTmp'
    /usr/bin/make  -f CMakeFiles/cmTC_7b38f.dir/build.make CMakeFiles/cmTC_7b38f.dir/build
    make[4]: Entering directory '/home/ubuntu/jedi/jedi-stack/pkg/nceplibs-develop/build/gfsio/src/gfsio-build/CMakeFiles/CMakeTmp'
    Building Fortran object CMakeFiles/cmTC_7b38f.dir/testFortranCompiler.f.o
    /optjedi/intel/compilers_and_libraries_2020.0.166/linux/mpi/intel64/bin/mpiifort    -c /home/ubuntu/jedi/jedi-stack/pkg/nceplibs-develop/build/gfsio/src/gfsio-build/CMakeFiles/CMakeTmp/testFortranCompiler.f -o CMakeFiles/cmTC_7b38f.dir/testFortranCompiler.f.o
    /optjedi/intel/compilers_and_libraries_2020.0.166/linux/mpi/intel64/bin/mpiifort: 1: eval: ifort: not found
    CMakeFiles/cmTC_7b38f.dir/build.make:85: recipe for target 'CMakeFiles/cmTC_7b38f.dir/testFortranCompiler.f.o' failed
    make[4]: *** [CMakeFiles/cmTC_7b38f.dir/testFortranCompiler.f.o] Error 127
    make[4]: Leaving directory '/home/ubuntu/jedi/jedi-stack/pkg/nceplibs-develop/build/gfsio/src/gfsio-build/CMakeFiles/CMakeTmp'
    Makefile:141: recipe for target 'cmTC_7b38f/fast' failed
    make[3]: *** [cmTC_7b38f/fast] Error 2
    make[3]: Leaving directory '/home/ubuntu/jedi/jedi-stack/pkg/nceplibs-develop/build/gfsio/src/gfsio-build/CMakeFiles/CMakeTmp'
```

**Comment by aerorahul** [comment]
> I'm also having trouble on AWS with intel. Strange. At first Cmake seems to find the compiler fine:
> 
> ```shell
> + cmake -DCMAKE_INSTALL_PREFIX=/optjedi/modules/intel-20.0.166/impi-20.0.166 -DDEPLOY=ON -DBUILD_CRTM=OFF -DBUILD_POST=OFF ..
> -- The C compiler identification is Intel 19.1.0.20191121
> -- The Fortran compiler identification is Intel 19.1.0.20191121
> ```
> 
> But then during the actual make it gets confused:
> 
> ```shell
> + sudo VERBOSE=Y make -j4
> [...]
> Re-run cmake no build system arguments
> [ 14%] Performing configure step for 'bacio'
> cd /home/ubuntu/jedi/jedi-stack/pkg/nceplibs-develop/build/bacio/src/bacio-build && /optjedi/modules/core/cmake/3.17.3/bin/cmake "-DCMAKE_PREFIX_PATH=/optjedi/modules/intel-20.0.166/impi-20.0.166;/optjedi/modules/intel-20.0.166/impi-20.0.166/bacio;/optjedi/modules/intel-20.0.166/impi-20.0.166/sigio;/optjedi/modules/intel-20.0.166/impi-20.0.166/sfcio;/optjedi/modules/intel-20.0.166/impi-20.0.166/gfsio;/optjedi/modules/intel-20.0.166/impi-20.0.166/ip;/optjedi/modules/intel-20.0.166/impi-20.0.166/sp;/optjedi/modules/intel-20.0.166/impi-20.0.166/w3nco;/optjedi/modules/intel-20.0.166/impi-20.0.166/g2;/optjedi/modules/intel-20.0.166/impi-20.0.166/g2tmpl;/optjedi/modules/intel-20.0.166/impi-20.0.166/nemsio;/optjedi/modules/intel-20.0.166/impi-20.0.166/nemsiogfs;/optjedi/modules/intel-20.0.166/impi-20.0.166/landsfcutil;/optjedi/modules/intel-20.0.166/impi-20.0.166/w3emc;/optjedi/modules/intel-20.0.166/impi-20.0.166/wrf_io;/optjedi/modules/intel-20.0.166/impi-20.0.166/bufr" -DCMAKE_INSTALL_PREFIX=/optjedi/modules/intel-20.0.166/impi-20.0.166/bacio-2.4.0 -DENABLE_TESTS=OFF -DCMAKE_C_COMPILER=/optjedi/intel/compilers_and_libraries_2020.0.166/linux/mpi/intel64/bin/mpiicc -DCMAKE_Fortran_COMPILER=/optjedi/intel/compilers_and_libraries_2020.0.166/linux/mpi/intel64/bin/mpiifort "-GUnix Makefiles" /home/ubuntu/jedi/jedi-stack/pkg/nceplibs-develop/build/bacio/src/bacio
> Re-run cmake no build system arguments
> -- The Fortran compiler identification is unknown
> /optjedi/intel/compilers_and_libraries_2020.0.166/linux/mpi/intel64/bin/mpiifort: 1: eval: ifort: not found
> -- Check for working Fortran compiler: /optjedi/intel/compilers_and_libraries_2020.0.166/linux/mpi/intel64/bin/mpiifort
> -- The Fortran compiler identification is unknown
> /optjedi/intel/compilers_and_libraries_2020.0.166/linux/mpi/intel64/bin/mpiifort: 1: eval: ifort: not found
> -- Check for working Fortran compiler: /optjedi/intel/compilers_and_libraries_2020.0.166/linux/mpi/intel64/bin/mpiifort
> -- Check for working Fortran compiler: /optjedi/intel/compilers_and_libraries_2020.0.166/linux/mpi/intel64/bin/mpiifort - broken
> CMake Error at /optjedi/modules/core/cmake/3.17.3/share/cmake-3.17/Modules/CMakeTestFortranCompiler.cmake:45 (message):
>   The Fortran compiler
> 
>     "/optjedi/intel/compilers_and_libraries_2020.0.166/linux/mpi/intel64/bin/mpiifort"
> 
>   is not able to compile a simple test program.
> 
>   It fails with the following output:
> 
>     Change Dir: /home/ubuntu/jedi/jedi-stack/pkg/nceplibs-develop/build/gfsio/src/gfsio-build/CMakeFiles/CMakeTmp
>     
>     Run Build Command(s):/usr/bin/make cmTC_7b38f/fast && make[3]: Entering directory '/home/ubuntu/jedi/jedi-stack/pkg/nceplibs-develop/build/gfsio/src/gfsio-build/CMakeFiles/CMakeTmp'
>     /usr/bin/make  -f CMakeFiles/cmTC_7b38f.dir/build.make CMakeFiles/cmTC_7b38f.dir/build
>     make[4]: Entering directory '/home/ubuntu/jedi/jedi-stack/pkg/nceplibs-develop/build/gfsio/src/gfsio-build/CMakeFiles/CMakeTmp'
>     Building Fortran object CMakeFiles/cmTC_7b38f.dir/testFortranCompiler.f.o
>     /optjedi/intel/compilers_and_libraries_2020.0.166/linux/mpi/intel64/bin/mpiifort    -c /home/ubuntu/jedi/jedi-stack/pkg/nceplibs-develop/build/gfsio/src/gfsio-build/CMakeFiles/CMakeTmp/testFortranCompiler.f -o CMakeFiles/cmTC_7b38f.dir/testFortranCompiler.f.o
>     /optjedi/intel/compilers_and_libraries_2020.0.166/linux/mpi/intel64/bin/mpiifort: 1: eval: ifort: not found
>     CMakeFiles/cmTC_7b38f.dir/build.make:85: recipe for target 'CMakeFiles/cmTC_7b38f.dir/testFortranCompiler.f.o' failed
>     make[4]: *** [CMakeFiles/cmTC_7b38f.dir/testFortranCompiler.f.o] Error 127
>     make[4]: Leaving directory '/home/ubuntu/jedi/jedi-stack/pkg/nceplibs-develop/build/gfsio/src/gfsio-build/CMakeFiles/CMakeTmp'
>     Makefile:141: recipe for target 'cmTC_7b38f/fast' failed
>     make[3]: *** [cmTC_7b38f/fast] Error 2
>     make[3]: Leaving directory '/home/ubuntu/jedi/jedi-stack/pkg/nceplibs-develop/build/gfsio/src/gfsio-build/CMakeFiles/CMakeTmp'
> ```

I think I had seen that issue, and so I pass CMAKE_C_COMPILER as an argument to each of the individual library projects.

Rather than @mmiesch going through these issues,  I am going to close this PR. I don't want @mmiesch to spend time on this.
I will re-open when it is good and ready.

**Comment by mmiesch** [comment]
Thanks @aerorahul !

**Pull request closed without merging**